### ISSUE TITLE
luajit support ?

### ISSUE THREAD

**Issue by mcarbonneaux** [haproxy/haproxy#915]
## What should haproxy do differently? 

- the luajit are more performante than lua 5.4 (that is allready more performante all lua precedante version)....
https://luajit.org/performance_x86.html
https://eklausmeier.wordpress.com/2020/05/14/performance-comparison-pallene-vs-lua-5-1-5-2-5-3-5-4-vs-c/

- more ligth in memory foot print...
- native support of ffi.

**Comment by wlallemand** [comment]
We already talk about this a long time ago, If I remember correctly the problem was that HAProxy require at least lua 5.3 and luajit only implements the 5.1 version. One of the problem was the yield implementation which is a requirement in HAProxy. Maybe it's possible to do this now with luajit, we should take a look at this.

**Comment by wtarreau** [comment]
Yep that's it. LuaJit seems very well suited for games and such well-controlled environments where performance is very important, but not much for anything subject to potentially hostile activity, especially since being limited to Lua 5.1, it cannot yield after a predefined number of instructions!

I'd say that most Lua users don't seek performance here but flexibility. For now one single thread runs Lua at a time and even with this limitation, almost nobody complains. Lua is mostly used for small glue that cannot easily be expressed using sample expressions, where the processing time is not significant to make a measurable difference, or conversely for large application components (authentication etc) where what matters is mostly the ability to integrate well into an existing architecture rather than the raw performance. Typically an authentication server only verifies auth for the first request and is not sollicited after that.

**Comment by mcarbonneaux** [comment]
i'm evaluating open source alternative for migration from proprietary lb that make possible the use of scripting extensivly to manipulate traffic.
is why i'm interogate me on the performance of the scripting language of using lua vs luajit...

**Comment by wtarreau** [comment]
If you need to manipulate traffic, you really want to be absolutely certain you can't loop forever in a script by accident, something that luajit cannot offer but lua does. However if you need to do a lot of things on traffic, there are filters in haproxy. They're used for compression or caching for example. You could very well write your own filter to perform your processing using native code. You could even link your code with luajit if that would help you port existing code. Look for `filter_kws` in flt_http_comp.c to get an idea. It's also possible to use SPOE for this, and place the processing on external nodes, but at the moment SPOE is limited to processing a single buffer.

**Comment by kanashimia** [comment]
Hello, I know this is an old thread, but I took a look at it recently,

> especially since being limited to Lua 5.1, it cannot yield after a predefined number of instructions!

Yielding after predefined number of instructions is done using lua_sethook, which exists in Lua 5.1 and in LuaJIT.

But problem is that HAProxy uses lua_yieldk (introduced in 5.2) very heavily, for yieldable C functions, so supporting LuaJIT as is isn't possible, will require quite a big code restructuring in a not a good way. And a similar related issue asking for lua_pcallk in LuaJIT was closed as WONTFIX. From my POV it is not worth supporting that.

Lua 5.2 should be possible to support, but I don't think there are any use cases for that.

What I found though, is that Luau actually has needed features, but the API is slightly different, it has luaL_callyieldable for calling yieldable C functions, and has interrupt hook / debuginterrupt hook + singlestep mode, which could be used instead of lua_sethook.
The API is slightly different, but theoretically Luau should be possible to support.
Rust mlua bindings had a PR that added yielding support and it had implementations for both Lua 5.2+ with lua_yieldk and for Luau with luaL_callyieldable.
I will look into maybe adding its support in haproxy next.
